// beachball/lib/changelog/mergeChangelogs.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.mergeChangelogs = void 0;
const generateTag_1 = require("../git/generateTag");
/**
 * Merge multiple package changelogs into one.
 * `name` and `version` will use the values from `primaryPackage`'s changelog.
 * `comments` are merged. `date` will be now.
 */
function mergeChangelogs(changelogs, primaryPackage) {
    var _a;
    if (changelogs.length < 1 || !primaryPackage) {
        return undefined;
    }
    const result = {
        name: primaryPackage.name,
        version: primaryPackage.version,
        tag: (0, generateTag_1.generateTag)(primaryPackage.name, primaryPackage.version),
        date: new Date(),
        comments: {},
    };
    for (const changelog of changelogs) {
        for (const changeType of Object.keys(changelog.comments)) {
            const comments = changelog.comments[changeType];
            if (comments?.length) {
                ((_a = result.comments)[changeType] ?? (_a[changeType] = [])).push(...comments);
            }
        }
    }
    return result;
}
exports.mergeChangelogs = mergeChangelogs;
