// beachball/lib/options/getDefaultOptions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getDefaultOptions = void 0;
const env_1 = require("../env");
/**
 * Default options.
 */
function getDefaultOptions() {
    return {
        access: 'restricted',
        all: false,
        authType: 'authtoken',
        branch: 'origin/master',
        bump: true,
        bumpDeps: true,
        canaryName: undefined,
        changehint: 'Run "beachball change" to create a change file',
        changeDir: 'change',
        command: 'change',
        commit: true,
        defaultNpmTag: 'latest',
        depth: undefined,
        disallowedChangeTypes: null,
        fetch: true,
        generateChangelog: true,
        gitTags: true,
        gitTimeout: undefined,
        message: '',
        new: false,
        path: '',
        publish: true,
        push: true,
        registry: 'https://registry.npmjs.org/',
        retries: 3,
        scope: null,
        tag: '',
        timeout: undefined,
        type: null,
        version: false,
        yes: env_1.env.isCI,
    };
}
exports.getDefaultOptions = getDefaultOptions;
